File: apps/command/lib/api/onboard.ts
/**
 * City-in-a-box jobs (`POST /v1/onboard`, `GET /v1/onboard/{id}`; CLAUDE.md 7.9, 12).
 *
 * The wizard polls rather than relying only on the WebSocket. A build is minutes long and the
 * one thing that must not happen on stage is a progress bar that stops moving because a socket
 * dropped; polling every second costs one small request and is the state of record.
 */

import { apiUrl } from "@/lib/api/client";

export type OnboardStatus = "none" | "queued" | "running" | "finished" | "failed";

/** The six steps the wizard shows, in order (CLAUDE.md 7.9). */
export type OnboardStepId =
  | "choose_area"
  | "fetch_open_data"
  | "condition_terrain"
  | "infer_drains"
  | "build_graph"
  | "first_forecast";

export interface OnboardJob {
  jobId: string | null;
  city: string;
  status: OnboardStatus;
  step: OnboardStepId;
  /** 0 to 1 across the whole build. */
  progress: number;
  startedAt: string | null;
  finishedAt: string | null;
  elapsedS: number;
  /** The pipeline's own log lines, newest last. Never composed here. */
  logTail: string[];
  firstRunId: string | null;
  error: string | null;
  /** Whether `city/<city>/segments.parquet` already exists, so a built city reads as built. */
  built: boolean;
}

function parse(body: Record<string, unknown>): OnboardJob {
  return {
    jobId: (body.job_id as string | null) ?? null,
    city: String(body.city ?? ""),
    status: (body.status as OnboardStatus) ?? "none",
    step: (body.step as OnboardStepId) ?? "choose_area",
    progress: Number(body.progress ?? 0),
    startedAt: (body.started_at as string | null) ?? null,
    finishedAt: (body.finished_at as string | null) ?? null,
    elapsedS: Number(body.elapsed_s ?? 0),
    logTail: (body.log_tail as string[]) ?? [],
    firstRunId: (body.first_run_id as string | null) ?? null,
    error: (body.error as string | null) ?? null,
    built: Boolean(body.built),
  };
}

async function call(path: string, init?: RequestInit): Promise<OnboardJob> {
  const response = await fetch(apiUrl(path), init);
  const body = (await response.json()) as Record<string, unknown>;
  if (!response.ok) {
    const envelope = body.error as { message?: string } | undefined;
    throw new Error(envelope?.message ?? `Onboarding failed: HTTP ${response.status}`);
  }
  return parse(body);
}

/** Start a build, or rejoin the one already running for this city. */
export function startOnboard(city: string, designStorm = "CHN-IDF-25yr"): Promise<OnboardJob> {
  return call("/v1/onboard", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ city, design_storm: designStorm, from_cache_only: true }),
  });
}

export function pollOnboard(jobId: string, signal?: AbortSignal): Promise<OnboardJob> {
  return call(`/v1/onboard/${encodeURIComponent(jobId)}`, { signal });
}

/** What the wizard asks on load, so a reopened tab rejoins a build already in flight. */
export function latestOnboard(city: string, signal?: AbortSignal): Promise<OnboardJob> {
  return call(`/v1/onboard/city/${encodeURIComponent(city)}`, { signal });
}

File: services/api/varuna_api/routers/onboard.py
"""City-in-a-box on stage (CLAUDE.md 7.9, 12; tasks P9.5, P9.6).

`POST /v1/onboard` starts a build in a background thread and answers immediately with a job; the
wizard polls `GET /v1/onboard/{job_id}` and also receives the pipeline's own `onboard.progress`
events over the WebSocket. Both carry the same state, so a dropped socket degrades to polling
rather than to a frozen progress bar.
"""

from __future__ import annotations

from pathlib import Path
from typing import Any

import structlog
from fastapi import APIRouter
from varuna_schemas.paths import city_dir

from varuna_api.onboard import get_job, latest_job, start_job
from varuna_api.routers.stubs import OnboardRequest
from varuna_api.state import api_error

log = structlog.get_logger("varuna.api.onboard")

router = APIRouter(prefix="/v1", tags=["onboard"])


@router.post("/onboard", status_code=202, summary="Start a city-in-a-box job")
def onboard(body: OnboardRequest) -> dict[str, Any]:
    """Build a city and run its first forecast.

    Validated through `OnboardRequest`, the documented shape (CLAUDE.md 12), so a malformed body
    gets the 422 envelope rather than a message about the city not existing.

    Answers 202 with the job straight away: the build is minutes of work and holding the request
    open for it would time out behind any proxy, and the wizard wants to render its first step
    immediately anyway.

    A request for a city that is already building returns that job rather than starting a second
    one - on stage, a double-click on "Start" must not raise a dialog.
    """
    from varuna_city.config import load_city_config
    from varuna_schemas.settings import get_settings

    # Refused before anything is validated or started. On 13 September 2026 a Chennai build was
    # started against the public Railway API: it downloaded tiles and OSM responses onto the
    # 500 MB volume beside two built cities, failed at `export` with ENOSPC, and left the volume
    # at 495 MB, after which the API crashed on its next restart and stayed down. The wizard is a
    # demo-laptop feature (CLAUDE.md 7.9 "on the demo laptop"), so the deployment says so.
    if not get_settings().varuna_onboard_enabled:
        raise api_error(
            403,
            "onboard_disabled",
            "City-in-a-box is switched off on this public deployment. A Chennai build needs about "
            "320 MB of terrain tiles, OSM responses and outputs, and the volume here is 500 MB "
            "shared with the Mumbai demo. Run the wizard against a local API (make demo), or "
            "build from the command line with make city CITY=chennai.",
        )

    city = body.city or "chennai"
    try:
        load_city_config(city)
    except (FileNotFoundError, KeyError, ValueError) as error:
        raise api_error(
            404,
            "unknown_city",
            f"No city config for {city!r}. Add services/city/configs/{city}.yaml.",
        ) from error

    state = start_job(
        city=city,
        design_storm=body.design_storm,
        from_cache_only=body.from_cache_only,
    )
    return state.to_dict()


@router.get("/onboard/{job_id}", summary="Job status, with the pipeline's own log tail")
def onboard_job(job_id: str) -> dict[str, Any]:
    """Poll one job. ``latest`` in place of an id returns the most recent job for a city."""
    state = get_job(job_id)
    if state is None:
        raise api_error(
            404, "no_job", f"No onboarding job {job_id!r}. Start one with POST /v1/onboard."
        )
    return state.to_dict()


@router.get("/onboard/city/{city}", summary="The most recent job for a city")
def onboard_latest(city: str) -> dict[str, Any]:
    """What the wizard asks on load, so a reopened tab rejoins a build already running."""
    state = latest_job(city)
    if state is None:
        return {
            "job_id": None,
            "city": city,
            "status": "none",
            "built": (Path(city_dir(city)) / "segments.parquet").is_file(),
        }
    payload = state.to_dict()
    payload["built"] = (Path(city_dir(city)) / "segments.parquet").is_file()
    return payload
